ui/components/app/currency-input/currency-input.js

```diff
@@ -41,84 +45,93 @@ export default function CurrencyInput({
   onPreferenceToggle,
   swapIcon,
   className = '',
+  // if null, the asset is the native currency
+  asset,
 }) {
   const t = useContext(I18nContext);
 
+  const assetDecimals = asset?.decimals || NATIVE_CURRENCY_DECIMALS;
+
   const preferredCurrency = useSelector(getNativeCurrency);
   const secondaryCurrency = useSelector(getCurrentCurrency);
-  const conversionRate = useSelector(getConversionRate);
+
+  const primarySuffix =
+    asset?.symbol || preferredCurrency || EtherDenomination.ETH;
+  const secondarySuffix = secondaryCurrency.toUpperCase();
+  const isLongSymbol = (primarySuffix?.length || 0) > LARGE_SYMBOL_LENGTH;
+
+  const [shouldDisplayFiat, setShouldDisplayFiat] = useState(featureSecondary);
   const showFiat = useSelector(getShouldShowFiat);
+  const hideSecondary = !showFiat;
+  const shouldUseFiat = hideSecondary ? false : Boolean(shouldDisplayFiat);
+  const isTokenPrimary = !shouldUseFiat;
+
+  const [tokenDecimalValue, setTokenDecimalValue] = useState('0');
+  const [fiatDecimalValue, setFiatDecimalValue] = useState('0');
+
   const chainId = useSelector(getCurrentChainId);
   const { ticker, type } = useSelector(getProviderConfig);
   const isOriginalNativeSymbol = useIsOriginalNativeTokenSymbol(
     chainId,
     ticker,
     type,
   );
-  const hideSecondary = !showFiat;
-  const primarySuffix = preferredCurrency || EtherDenomination.ETH;
-  const secondarySuffix = secondaryCurrency.toUpperCase();
-
-  const [isSwapped, setSwapped] = useState(false);
-  const [newHexValue, setNewHexValue] = useState(hexValue);
-  const [shouldDisplayFiat, setShouldDisplayFiat] = useState(featureSecondary);
-  const shouldUseFiat = hideSecondary ? false : Boolean(shouldDisplayFiat);
 
-  const getDecimalValue = () => {
-    const decimalValueString = shouldUseFiat
-      ? getValueFromWeiHex({
-          value: hexValue,
-          toCurrency: secondaryCurrency,
-          conversionRate,
-          numberOfDecimals: 2,
-        })
-      : getValueFromWeiHex({
-          value: hexValue,
-          toCurrency: EtherDenomination.ETH,
-          numberOfDecimals: 8,
-        });
-
-    return Number(decimalValueString) || 0;
-  };
+  const tokenToFiatConversionRate = useTokenExchangeRate(asset?.address);
 
-  const initialDecimalValue = hexValue ? getDecimalValue() : 0;
+  const processNewDecimalValue = useProcessNewDecimalValue(
+    assetDecimals,
+    isTokenPrimary,
+    tokenToFiatConversionRate,
+  );
 
   const swap = async () => {
     await onPreferenceToggle();
-    setSwapped(!isSwapped);
     setShouldDisplayFiat(!shouldDisplayFiat);
   };
 
+  // if the conversion rate is undefined, do not allow a fiat input
+  useEffect(() => {
+    if (isTokenPrimary) {
+      return;
+    }
+
+    if (!tokenToFiatConversionRate) {
+      onPreferenceToggle();
+      setShouldDisplayFiat(false);
+    }
+  }, [tokenToFiatConversionRate, isTokenPrimary, onPreferenceToggle]);
+
```

Comment:
I think that's what this `useEffect` is doing, right?

Reply:
This useEffect toggles back to non-fiat after the user toggles to fiat and the conversion is not available. However, imo we shouldn't even allow the user to toggle to fiat if we know the conversion rate is undefined (`showFiat` reflects this state)

I recommend disabling or hiding the swap button if `!showFiat`. Otherwise, the screen may seem unresponsive (or flickery) to the user when they attempt to toggle to fiat and this useEffect flips it back to token. If we do this, the useEffect can be removed

Reply:
Good call, I added logic to remove the swap icon when it cannot be used anyways.

However, we still need the useEffect to ensure it doesn't get stuck in a fiat state.